backends/ignite.go

```diff
@@ -0,0 +1,256 @@
+// been previously stored under 'key' if its TTL is still current. We can tell when a key is not found
+// when Ignite doesn't return an error, nor a 'Status' different than zero, but the 'Response' field is
+// empty. Get can also return Ignite server-side errors
+func (ig *IgniteBackend) Get(ctx context.Context, key string) (string, error) {
+	urlCopy := *ig.serverURL
+	q := urlCopy.Query()
+	q.Set("cmd", "get")
+	q.Set("key", key)
+
+	urlCopy.RawQuery = q.Encode()
+
+	responseBytes, err := ig.sender.DoRequest(ctx, &urlCopy, ig.headers)
+	if err != nil {
+		return "", err
+	}
+
+	// Unmarshal response
+	igniteResponse := getResponse{}
+
+	if unmarshalErr := json.Unmarshal(responseBytes, &igniteResponse); unmarshalErr != nil {
+		return "", fmt.Errorf("Unmarshal response error: %s; Response body: %s", unmarshalErr.Error(), string(responseBytes))
+	}
+
+	// Validate response
+	if len(igniteResponse.Error) > 0 {
+		return "", utils.NewPBCError(utils.GET_INTERNAL_SERVER, igniteResponse.Error)
+	} else if igniteResponse.Status > 0 {
+		return "", utils.NewPBCError(utils.GET_INTERNAL_SERVER, "Ignite response.Status not zero")
+	} else if len(igniteResponse.Response) == 0 {
+		return "", utils.NewPBCError(utils.KEY_NOT_FOUND)
+	}
+
+	return igniteResponse.Response, nil
+}
+
+// putResponse is used to unmarshal the Ignite server's response to a PUT request with
+// the "cmd" URL query field set to "putifabs"
+type putResponse struct {
+	Error    string `json:"error"`
+	Response bool   `json:"response"`
+	Status   int    `json:"successStatus"`
+}
+
+// Put implements the Backend interface to comunicates with the Ignite storage service to perform
+// a "putifabs" command in order to store the "value" parameter only if the "key" doesn't exist in
+// the storage already. Returns RecordExistsError or whatever PUT_INTERNAL_SERVER error we might
+// find in the storage side
+func (ig *IgniteBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
+
+	urlCopy := *ig.serverURL
+	q := urlCopy.Query()
+	q.Set("cmd", "putifabs")
+	q.Set("key", key)
+	q.Set("val", value)
+	q.Set("exp", fmt.Sprintf("%d", ttlSeconds*1000))
+
+	urlCopy.RawQuery = q.Encode()
+
+	responseBytes, err := ig.sender.DoRequest(ctx, &urlCopy, ig.headers)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal response
+	igniteResponse := putResponse{}
+	if unmarshalErr := json.Unmarshal(responseBytes, &igniteResponse); unmarshalErr != nil {
+		return fmt.Errorf("Unmarshal response error: %s; Response body: %s", unmarshalErr.Error(), string(responseBytes))
+	}
+
+	// Validate response
+	if len(igniteResponse.Error) > 0 {
+		return utils.NewPBCError(utils.PUT_INTERNAL_SERVER, igniteResponse.Error)
+	}
+
+	if igniteResponse.Status > 0 {
+		return utils.NewPBCError(utils.PUT_INTERNAL_SERVER, "Ignite responded with non-zero successStatus code")
+	}
+
+	if !igniteResponse.Response {
```

Comment:
Is the `!` supposed to be here? If `igniteResponse.Response` is true a record already exists right?

Reply:
Given that we are not rewritting values for existing keys, we are using Apache Ignite's `putifabsent` command. According to the [put-if-absent docs](https://ignite.apache.org/docs/latest/restapi#put-if-absent): "The response field contains `true` if the entry was put, `false` otherwise."